### ISSUE TITLE
ElixirLS v0.17.1 in VSCode significantly changes token scopes, affecting syntax highlighting

### ISSUE THREAD

**Issue by bceskavich** [elixir-lsp/elixir-ls#1006]
### Environment

* Elixir & Erlang versions (elixir --version): Elixir v1.14.5, Erlang/OTP 25
* Elixir Language Server version: v0.17.1
* Operating system: macOS Ventura 13.2.1
* Editor or IDE name (e.g. Emacs/VSCode): VSCode
* Editor Plugin/LSP Client name and version: ElixirLS v0.17.1

### Current behavior

The latest release of ElixirLS for VSCode seems to remove a number of token scopes for Elixir, which in turn significantly changes the appearance of multiple syntax highlighting themes. The example below shows the following cases:

- Differentiation between module definition + module alias usage has disappeared
- Identification of module attributes as distinct entities has disappeared

I am the author of the [Dracula At Night](https://marketplace.visualstudio.com/items?itemName=bceskavich.theme-dracula-at-night) VSCode theme, and the examples below use this theme. However, I've verified that this issue affects other popular themes as well (e.g. [Cobalt2](https://marketplace.visualstudio.com/items?itemName=wesbos.theme-cobalt2)).

| <= v0.17.0 | v0.17.1 |
| - | - |
| <img width="707" alt="image" src="https://github.com/elixir-lsp/elixir-ls/assets/5049830/bc908603-ee9a-4875-9bef-7b90da05a7c0"> | <img width="707" alt="image" src="https://github.com/elixir-lsp/elixir-ls/assets/5049830/b926fcd2-64fe-4d7b-8a53-f097779b27d1"> |

### Expected behavior

Consistent token scopes and syntax highlighting across versions.

**Comment by lukaszsamson** [comment]
Yes, there has been some changes aiming to correct issues in the grammar

> Differentiation between module definition + module alias usage has disappeared

Should there be difference? In elixir AST both are the same.
note `{:__aliases__, [line: _], [:Other, :Asdf]}` in
```
iex(4)> Code.string_to_quoted("""
...(4)> defmodule Other.Asdf do
...(4)>   def a do
...(4)>     Other.Asdf.asd()
...(4)>   end
...(4)> end
...(4)> """)
{:ok,
 {:defmodule, [line: 1],
  [
    {:__aliases__, [line: 1], [:Other, :Asdf]},
    [
      do: {:def, [line: 2],
       [
         {:a, [line: 2], nil},
         [
           do: {{:., [line: 3],
             [{:__aliases__, [line: 3], [:Other, :Asdf]}, :asd]}, [line: 3], []}
         ]
       ]}
    ]
  ]}}
```
anyway, the difference is still there
usage in module is marked with scope `meta.module.elixir`
<img width="749" alt="Screenshot 2023-10-17 at 16 23 37" src="https://github.com/elixir-lsp/elixir-ls/assets/1078186/f8bd6923-c85a-4f24-b2ae-f2a11a75df90">
other usages does not have that scope
<img width="743" alt="Screenshot 2023-10-17 at 16 23 53" src="https://github.com/elixir-lsp/elixir-ls/assets/1078186/e864326a-71a1-45cf-9f95-a7d896ab7c0a">


> Identification of module attributes as distinct entities has disappeared

<img width="677" alt="Screenshot 2023-10-17 at 16 27 08" src="https://github.com/elixir-lsp/elixir-ls/assets/1078186/143a8f7b-1668-4b3b-ab14-47f778afb983">

Do you mean that the scope has changed from `variable.other.readwrite.module.elixir` to more standard `variable.other.constant.elixir`? There were no `entity.name.` scope on module attributes.

> Expected behavior Consistent token scopes and syntax highlighting across versions.

I cannot evolve the grammar without making some changes

**Comment by bceskavich** [comment]
Thanks for the fast response @lukaszsamson!

Also, thanks for clarifying both of the items I highlighted. The decision to highlight module alias definitions vs usages differently is definitely a choice I've made with my theme, however one that is a common across other themes. Regarding the module attribute change, I seem to have misread the change in scopes and see that there's still a distinction there I can use to highlight those attributes as desired.

Lastly, while I understand the desire to make the grammar here more correct, to me these updates feel like a breaking change that went [unannounced in a patch version release](https://github.com/elixir-lsp/elixir-ls/releases/tag/v0.17.1). At the very least it would be helpful in the future if such changes were documented and part of a minor (not patch) release. 

In addition, I would advocate for considering significant changes in the token identification grammar as a breaking change in the future. There are thousands of users of this VSCode plugin who are suddenly experiencing a drastic change in appearance of Elixir code across multiple themes as a result of this update. And for those who aren't theme maintainers they may be searching in confusion to understand why this happened (i.e. is this the theme they use? is it their VSCode version? is it the LS plugin?).

**Comment by lukaszsamson** [comment]
The question now is if the current rules allow you to make your theme work as you wish or do you need some changes? Which areas could be improved?

> Lastly, while I understand the desire to make the grammar here more correct, to me these updates feel like a breaking change that went [unannounced in a patch version release](https://github.com/elixir-lsp/elixir-ls/releases/tag/v0.17.1). At the very least it would be helpful in the future if such changes were documented and part of a minor (not patch) release.

I understand that it can be frustrating, especially given that some of the changes was unintended (https://github.com/elixir-lsp/vscode-elixir-ls/issues/382). Grammar changes was mentioned in v0.17.0 and v0.17.1 release notes
https://github.com/elixir-lsp/vscode-elixir-ls/blob/master/CHANGELOG.md. Besides 0.x in SemVer does not guarantee no breaking changes.

> In addition, I would advocate for considering significant changes in the token identification grammar as a breaking change in the future

Thanks. I now know that more cere is needed here. I'm afraid there will be more changes in the future. Elixir is notoriously difficult to parse with textmate grammars and LSP symantic tokens seem to be the solution
